relentless/optimize/method.py

```diff
@@ -121,11 +131,144 @@ def abs_tol(self, value):
         else:
             self._abs_tol = abs_tol
 
+class LineSearch:
+    r"""Line search algorithm.
+
+    For an :class:`~relentless.optimize.ObjectiveFunction` :math:`f\left(\mathbf{x}\right)`,
+    the line search algorithm seeks to take a single, optimally-sized step towards
+    the minimum of the function. (Does not necessarily converge to minimum.)
+
+    A search interval :math:`\mathbf{d}` is specified, and normalized to :math:`\hat{\mathbf{d}}`:
+
+    .. math::
+
+        \mathbf{d} = \mathbf{x}_{end} - \mathbf{x}_{start}
+
+        \hat{\mathbf{d}} = \frac{\mathbf{d}}{\lVert\mathbf{d}\rVert}
+
+    Then, a 'target' value :math:`t` is defined as:
+
+    .. math::
+
+        t = -\hat{\mathbf{d}} \cdot \nabla{f\left(\mathbf{x}\right)}
+
+    The target at the start of the search interval is always negative. If the
```

Comment:
Wait, I've got myself confused now... if d points from start to end, then for steepest descent, it points in the direction of -grad(f). This would mean t > 0 at the start of the search interval?

I feel like these signs don't match the code below.

Reply:
Yeah, Mike is correct. d is a descent direction, so the way we have defined t, t > 0 means we decrease the function by moving in the direction of d. t < 0 means we increase the function by moving opposite the direction of d. t will always be positive at the start point (if it wasn't d would be a very lousy choice for a search direction!)